include/exadg/incompressible_navier_stokes/spatial_discretization/spatial_operator_base.cpp

```diff
@@ -249,20 +249,87 @@ template<int dim, typename Number>
 void
 SpatialOperatorBase<dim, Number>::distribute_dofs()
 {
+  if(param.spatial_discretization == SpatialDiscretization::L2)
+  {
+    fe_u = std::make_shared<dealii::FESystem<dim>>(dealii::FE_DGQ<dim>(param.degree_u), dim);
+
+    pcout << std::endl
+          << "Discontinuous Galerkin finite element discretization:" << std::endl
+          << std::endl
+          << std::flush;
+  }
+  else if(param.spatial_discretization == SpatialDiscretization::HDIV)
+  {
+    // In-parameter for RT is the degree in tangent direction, i.e the degree in normal direction
+    // will be degree+1. The FE also saves fe_u.degree = degree+1.
+    // degree - 1 is used so that the same Gauss quadrature can be used for exact integration.
+    fe_u = std::make_shared<dealii::FE_RaviartThomasNodal<dim>>(param.degree_u - 1);
+
+    pcout << std::endl
+          << "Hdiv-conforming Raviart-Thomas finite element discretization:" << std::endl
+          << "(Note that the given degree for the velocity is the degree in normal direction."
+          << std::endl
+          << " The degree in tangent direction is degree - 1. Note also that the number of dofs"
+          << std::endl
+          << "given by the dof_handler does not take the constraint of periodic boundaries into"
+          << std::endl
+          << " account.)" << std::endl
+          << std::endl
+          << std::flush;
+  }
+  else
+    AssertThrow(false, dealii::ExcMessage("FE not implemented."));
+
   // enumerate degrees of freedom
   dof_handler_u.distribute_dofs(*fe_u);
   dof_handler_p.distribute_dofs(fe_p);
   dof_handler_u_scalar.distribute_dofs(fe_u_scalar);
 
+  if(param.spatial_discretization == SpatialDiscretization::HDIV)
+  {
+    // We need to make sure the normal dofs are shared between cells on the periodic boundaries,
+    // since these are contiguous for HDIV.
+    dealii::IndexSet relevant_dofs;
+    dealii::DoFTools::extract_locally_relevant_dofs(dof_handler_u, relevant_dofs);
+    constraint_u.reinit(relevant_dofs);
+
+    auto const & periodic_face_pair = grid->periodic_faces;
+    for(auto const & face : periodic_face_pair)
+      dealii::DoFTools::make_periodicity_constraints(
+        dof_handler_u,
+        face.cell[0]->face(face.face_idx[0])->boundary_id(),
+        face.cell[1]->face(face.face_idx[1])->boundary_id(),
+        face.face_idx[0] / 2,
+        constraint_u);
+
+
+    // Set boundaries constraints
+    if(boundary_descriptor->velocity->symmetry_bc.empty() == false)
+    {
+      for(auto bc : boundary_descriptor->velocity->symmetry_bc)
+        dealii::VectorTools::project_boundary_values_div_conforming(
```

Comment:
Alternatively, these rows could be commented out for now.

Reply:
do symmetry boundary conditions work now?

Reply:
Yes!